apps/web-portal-client/lib/hooks/useFetchData.js

```diff
@@ -0,0 +1,46 @@
+import {useEffect, useReducer, useRef} from 'react'
+
+export function useFetchData(callback) {
+  const fetchDataReducer = (state, action) => {
+    switch (action.type) {
+      case 'FETCH_INIT': {
+        return {...state, isLoading: true, isError: false}
+      }
+      case 'FETCH_SUCCESS': {
+        return {...state, isLoading: false, isError: false, data: action.payload}
+      }
+      case 'FETCH_FAILURE': {
+        return {...state, isLoading: false, isError: action.payload}
+      }
+      default:
+        return state
+    }
+  }
+
+  const [state, dispatch] = useReducer(fetchDataReducer, {
+    data: undefined,
+    isLoading: false,
+    isError: false,
+  })
+
+  const isMounted = useRef(false)
+
+  useEffect(() => {
+    isMounted.current = true
+    return () => (isMounted.current = false)
+  }, [])
+
+  function doFetch() {
+    dispatch({type: 'FETCH_INIT'})
+
+    callback()
+      .then((result) => {
+        if (isMounted.current) dispatch({type: 'FETCH_SUCCESS', payload: result})
```

Comment:
it works well, but we also should add API request cancellation on unmount
![image](https://user-images.githubusercontent.com/13466225/111074773-8c830380-84dc-11eb-9b4c-3534012e2e7c.png)